safe_prime_generator.go

```diff
@@ -63,18 +63,19 @@ func GenerateSafePrime(
 	concurrencyLevel int,
 	timeout time.Duration,
 	random io.Reader,
-) (p *big.Int, q *big.Int, err error) {
+) (*big.Int, *big.Int, error) {
 	if bitLen < 6 {
 		return nil, nil, errors.New("safe prime size must be at least 6 bits")
 	}
 
 	primeChan := make(chan safePrime, concurrencyLevel)
 	errChan := make(chan error, concurrencyLevel)
 
+	waitGroup := &sync.WaitGroup{}
+
 	defer close(primeChan)
 	defer close(errChan)
-
-	waitGroup := &sync.WaitGroup{}
+	defer waitGroup.Wait()
```

Comment:
My personal opinion is that we should stop waiting as soon as we fire off the last goroutine (which is the timeout goroutine). Is there a specific advantage we get from deferring this this?

Reply:
So right here:
```
	ctx, cancel := context.WithCancel(context.Background())

	for i := 0; i < concurrencyLevel; i++ {
		waitGroup.Add(1)
		runGenPrimeRoutine(
			ctx, primeChan, errChan, waitGroup, random, bitLen,
		)
	}

	// Cancel after the specified timeout.
	go func() {
		time.Sleep(timeout)
		// mutex.Lock()
		cancel()
		// mutex.Unlock()
	}()
	waitGroup.Wait()

	select {
	case result := <-primeChan:
```

Is what I have above still valid?

Reply:
Having the `waitGroup.Wait()` before the `select` means we'll need to wait for all goroutines to finish before we go to the `select`. And we don't want it :wink:

We want the first goroutine to win and cancel the rest, so the `waitGroup.Wait()` is just for graceful exiting. The ideal scenario is that the first goroutine finds a prime, write it to the result channel and then we `cancel()` the rest of them in the `select`. We do `waitGroup.Wait()` at the end to make sure they are all done.

Reply:
Oh! Cool, got it! Yep, the `cancel()` makes sense (which then triggers) the `Wait()` to execute via the `defer`.